tensorpipe/channel/cuda_xth/channel_impl.cc

```diff
@@ -287,7 +329,26 @@ void ChannelImpl::waitOnStartEventAndCopyAndSyncWithSourceStream(
 
   TP_VLOG(6) << "Channel " << id_ << " is copying payload (#"
              << op.sequenceNumber << ")";
-  op.process();
+  if (op.isSrcCudaBuffer && op.isCudaBuffer) {
+    CudaDeviceGuard guard(op.deviceIdx);
+    TP_CUDA_CHECK(cudaStreamWaitEvent(op.stream, op.startEv, 0));
+    TP_CUDA_CHECK(cudaMemcpyAsync(
+        op.ptr, op.srcPtr, op.length, cudaMemcpyDeviceToDevice, op.stream));
+
+    CudaEvent stopEv(op.deviceIdx);
+    stopEv.record(op.stream);
+    stopEv.wait(op.srcStream, op.srcDeviceIdx);
+  } else if (op.isSrcCudaBuffer && !op.isCudaBuffer) {
+    // TODO: This call blocks the host thread, defer it to an other thread.
+    TP_CUDA_CHECK(cudaMemcpyAsync(
+        op.ptr, op.srcPtr, op.length, cudaMemcpyDeviceToHost, op.srcStream));
+  } else if (!op.isSrcCudaBuffer && op.isCudaBuffer) {
+    TP_CUDA_CHECK(cudaMemcpyAsync(
+        op.ptr, op.srcPtr, op.length, cudaMemcpyHostToDevice, op.stream));
+    TP_CUDA_CHECK(cudaStreamSynchronize(op.stream));
```

Comment:
Why do we need this? Currently the `cudaMemcpyAsync` is blocking (even though you didn't put a comment in this case) hence it already in a sense synchronously waits for the operation to complete on the stream right?

Reply:
According to the CUDA docs it is not blocking:
> For all other transfers, the function is fully asynchronous. If pageable memory must first be staged to pinned memory, this will be handled asynchronously with a worker thread.

Reply:
Ok, I didn't know this, and I find it a bit surprising (why use a CUDA thread when we could do it ourselves?), but fine...